### PR TITLE
Updated Usage with working code

### PR DESCRIPTION
**Pull request by RichardLitt**
I was not able to get it working as it was; this should help a bit. See https://github.com/ipfs/interface-ipfs-core/issues/43. Also, updated ToC.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -75,22 +60,42 @@ In Go land:
...
+Install `interface-ipfs-core` as one of the dependencies of your project. Install the following test file:
 
-```
-var test = require('interface-ipfs-core')
+```js
+/* eslint-env mocha */
+'use strict'
+
+const test = require('interface-ipfs-core')
+const IPFS = require('ipfs')
```
Comment (dignifiedquire) [inline README.md]: it should be made clear that this does should be your own ipfs instance and not the package `ipfs`.

File: README.md [hunk 2]
```diff
@@ -75,22 +60,42 @@ In Go land:
...
+    })
   },
   teardown: function (cb) {
     cb()
   }
 }
 
-// use all of the test suits
-test.all(common)
+test.files(common)
```
Comment (dignifiedquire) [inline README.md]: This is only an example of running the `files` tests, not all of them, this should be noted with a reference to where one can find a list of all of them

File: package.json [hunk 3]
```diff
@@ -36,7 +36,8 @@
     "readable-stream": "1.1.13"
   },
   "devDependencies": {
-    "aegir": "^3.2.0"
+    "aegir": "^3.2.0",
+    "ipfs": "^0.14.1"
```
Comment (dignifiedquire) [inline package.json]: This is not a good idea, this adds the dependency to all the repos requiring it even though they might not need it

File: test.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    ipfs.load(() => {
+      cb(null, ipfs)
+    })
+  },
+  teardown: function (cb) {
+    cb()
+  }
+}
+
+test.files(common)
```
Comment (dignifiedquire) [inline test.js]: This should be in something like an `examples` folder or even in another repo, as this is only for show.

File: README.md (hunk 2 above)
Comment (RichardLitt) [inline README.md]: Where would that be?

File: test.js (hunk 4 above)
Comment (RichardLitt) [inline test.js]: Moved.

File: README.md (hunk 2 above)
Comment (dignifiedquire) [inline README.md]: here: https://github.com/ipfs/interface-ipfs-core/blob/master/src/index.js

File: README.md [hunk 5]
```diff
@@ -17,22 +18,6 @@
...
-    - [`object.put`](/API/object#objectput)
-    - [`object.get`](/API/object#objectget)
-    - [`object.data`](/API/object#objectdata)
-    - [`object.links`](/API/object#objectlinks)
-    - [`object.stat`](/API/object#objectstat)
-    - [`object.patch`](/API/object#objectpatch)
-      - [`object.patch.addLink`](/API/object#objectpatchaddlink)
-      - [`object.patch.rmLink`](/API/object#objectpatchrmlink)
-      - [`object.patch.appendData`](/API/object#objectpatchappenddata)
-      - [`object.patch.setData`](/API/object#objectpatchsetdata)
```
Comment (dignifiedquire) [inline README.md]: Why did you remove all of these?

File: README.md [hunk 6]
```diff
@@ -75,22 +60,42 @@ In Go land:
 
 ### Node.js
 
-Install `interface-ipfs-core` as one of the dependencies of your project and as a test file. Then, using `mocha` (for Node.js) or a test runner with compatible API, do:
+Install `interface-ipfs-core` as one of the dependencies of your project. Install the following test file:
 
-```
-var test = require('interface-ipfs-core')
+```js
+/* eslint-env mocha */
```
Comment (dignifiedquire) [inline README.md]: This is not needed, as neither `describe` nor `it` are used in this code snippet.

File: README.md [hunk 7]
```diff
@@ -75,22 +60,42 @@ In Go land:
...
+'use strict'
+
+const test = require('interface-ipfs-core')
+const IPFS = require('ipfs') // This should be your own instance, not the package ipfs.
```
Comment (dignifiedquire) [inline README.md]: Maybe sth like `const IPFS = getYourIPFSInstance()` instead?

File: README.md [hunk 8]
```diff
@@ -75,22 +60,42 @@ In Go land:
...
 
-var common = {
+const common = {
   setup: function (cb) {
-    cb(null, yourIPFSInstance)
+    const ipfs = new IPFS()
+    ipfs.load(() => {
```
Comment (dignifiedquire) [inline README.md]: `ipfs.load` is specific to `js-ipfs` so I am not sure it makes sense to have in here.

File: README.md [hunk 9]
```diff
@@ -75,22 +60,42 @@ In Go land:
...
 
-// use all of the test suits
-test.all(common)
+test.files(common)
+```
+
+Then, using `mocha` (for Node.js) or a test runner with compatible API, do:
+
+```sh
+mocha test.js
```
Comment (dignifiedquire) [inline README.md]: Caan you add `$` in front please, as well as below

File: examples/test.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+/* eslint-env mocha */
```
Comment (dignifiedquire) [inline examples/test.js]: Why the duplication of this file, with the one in readme?

File: README.md (hunk 9 above)
Comment (diasdavid) [inline README.md]: or a `>`

File: README.md (hunk 7 above)
Comment (diasdavid) [inline README.md]: 👍

File: README.md (hunk 5 above)
Comment (RichardLitt) [inline README.md]: They are not in the file. The Table of Contents should reflect the current file.

File: README.md (hunk 7 above)
Comment (RichardLitt) [inline README.md]: Can we make this process cleaner? How do I set up my own instance, easily?

File: README.md (hunk 8 above)
Comment (RichardLitt) [inline README.md]: Should I just remove it?

File: examples/test.js (hunk 10 above)
Comment (RichardLitt) [inline examples/test.js]: I want to be able to run it, immediately. You can't do that if it is stuck in a README.

File: README.md (hunk 5 above)
Comment (diasdavid) [inline README.md]: It is the table of contents of the repo, which includes the several APIs spec definitions. Otherwise how will devs be able to open the repo and jump quickly to where they need?

File: README.md (hunk 8 above)
Comment (diasdavid) [inline README.md]: yes please :)

File: examples/test.js (hunk 10 above)
Comment (diasdavid) [inline examples/test.js]: There isn't however (or will be) an implementation of IPFS inside this repo.

Comment (RichardLitt) [inline examples/test.js]: There still ought to be a way to run the code, shouldn't there? Otherwise it's not a usable example. See title of the PR.

**Comment by diasdavid** [comment]
@RichardLitt are you familiar with:

- https://github.com/maxogden/abstract-blob-store
- https://github.com/libp2p/interface-transport
- https://github.com/libp2p/interface-connection
- https://github.com/libp2p/interface-stream-muxer
- https://github.com/libp2p/interface-peer-routing
- https://github.com/libp2p/interface-record-store
- https://github.com/libp2p/interface-connection
- https://github.com/libp2p/interface-libp2p

These are all spec definitions plus tests that can be used to make sure that the implementations are compliant.

It would actually be good if https://github.com/ipfs/http-api-spec would have a battery of tests that could be used to test the API, so that we could run them against go-ipfs and js-ipfs. @whyrusleeping how interesting would it be to you to have the CLI extracted from go-ipfs, so that it can be compiled as a standalone binary so that we can put the shardness tests available to run against any IPFS implementation? (Might be better to talk about this in the http-api-spec repo)

**Comment by RichardLitt** [comment]
So it's common to not have runnable usage code? Can we compromise by adding a paragraph explaining out how to set up your own IPFS Instance that can be called programmatically? How would I do that?

**Comment by RichardLitt** [comment]
And yes, please raise an issue about that in http-api-spec. I've been wanting that for a long time, but I can't do it alone.

**Comment by diasdavid** [comment]
https://github.com/ipfs/interface-ipfs-core/pull/44#issuecomment-237845345

Yes, it is common not to have runnable code. Note that these tests are also ran in js-ipfs-api, which doesn't spawn the node at all.

https://github.com/ipfs/interface-ipfs-core/pull/44#issuecomment-237845417

There you go: https://github.com/ipfs/http-api-spec/issues/113

**Comment by RichardLitt** [comment]
Thanks. Ok, changed to yourIPFSInstance, removed example code. 

I still think we need a paragraph on how to spawn your own instance. Can you describe that to me, in simple writing, here?

**Comment by RichardLitt** [comment]
@diasdavid I'd like to figure how how to spawn an instance, so I can write a better doc here about how to do that. Where should I ask that? On IRC?

File: README.md [hunk 11]
```diff
@@ -76,22 +76,38 @@ In Go land:
...
-Install `interface-ipfs-core` as one of the dependencies of your project and as a test file. Then, using `mocha` (for Node.js) or a test runner with compatible API, do:
+Install `interface-ipfs-core` as one of the dependencies of your project. Install the following test file:
 
-```
-var test = require('interface-ipfs-core')
+```js
+'use strict'
+
+const test = require('interface-ipfs-core')
+const yourIPFSInstance = require('ipfs')
```
Comment (diasdavid) [inline README.md]: This doesn't yield an IPFS instance.

**Comment by diasdavid** [comment]
> @diasdavid I'd like to figure how to spawn an instance.

That is the thing, it is not up to the 'spec' to define how to spawn an IPFS instance. The tests only tell you that you have a 'correct' IPFS instance if the thing you pass to it is an IPFS instance that has all the functions and behaviour the tests expect.

**Comment by RichardLitt** [comment]
But you can't run the tests if you don't know how to run an instance. Isn't it important to tell people how to run the tests?

**Comment by diasdavid** [comment]
> But you can't run the tests if you don't know how to run an instance. Isn't it important to tell people how to run the tests?

Not quite, if you want to use this test suite to test 'your own IPFS interface implementation' and you don't know how to create an instance from your implementation, you are in a deeper problem.

Note that this tests get executed in js-ipfs and js-ipfs-api repos, as they are required (which is why the tests are in the src folder, because this is a test suite module). 

Hope this clarifies, closing this PR.
